### PR TITLE
QF - add interface to set device user function

### PR DESCRIPTION
**Pull request by jeremylt**
@YohannDudouit @pazner @jedbrown 

This is half of the interface discussed on Slack. The other half would be the actual backend impl.

ToDo:
- [x] Add backend impl for CUDA backends (and HIP if we can)
- [ ] Decide how to test (HIP CI easier than CUDA right now)

### CODE REVIEW

File: include/ceed-cuda.h
Comment (jeremylt) [reply include/ceed-cuda.h:23]: You're right on both counts. I'll get it in the morning with other edits.

**Comment by jeremylt** [comment]
Updated based upon feedback. What do we think now?

**Comment by jeremylt** [comment]
@pazner if you're happy with your modification to the CUDA backends to accept CUFunctions, we can add that to this PR (do you are the right permissions or do we need to ask Jed?)

File: interface/ceed-qfunction.c
Comment (pazner) [inline interface/ceed-qfunction.c:648]:
```suggestion
    CeedDebug("Backend does not support CUfunction pointers for QFunctions.");
```
Is this right?

File: interface/ceed-qfunction.c
Comment (pazner) [inline interface/ceed-qfunction.c:673]: Same as above?

File: backends/cuda/ceed-cuda-qfunction.c
Comment (pazner) [inline backends/cuda/ceed-cuda-qfunction.c:169]: Did these lines serve a purpose?

File: interface/ceed-qfunction.c
Comment (jeremylt) [reply interface/ceed-qfunction.c:673]: yes, I believe that you are correct

File: backends/cuda/ceed-cuda-qfunction.c
Comment (jeremylt) [reply backends/cuda/ceed-cuda-qfunction.c:169]: I think those predate single source Qfunctions and were missed in the transition

**Comment by pazner** [comment]
> @pazner if you're happy with your modification to the CUDA backends to accept CUFunctions, we can add that to this PR (do you are the right permissions or do we need to ask Jed?)

Sure, I just pushed my changes. It uses an empty string for the source path to indicate that the user should supply the Q-function themselves. There is no real error checking if they give an empty string and don't supply the CUfunction (it will just crash).

File: backends/cuda/ceed-cuda-qfunction-load.cpp
Comment (jeremylt) [inline backends/cuda/ceed-cuda-qfunction-load.cpp:56]:
```suggestion
  if (!data->QFunctionSource)
    return CeedError(ceed, "No QFunction source or CUfunction provided.");
```

would require moving up lines 128,129

**Comment by jeremylt** [comment]
Thanks. I'll fix the Debug in the morning and then I think we're set.

**Comment by jeremylt** [comment]
@jedbrown @pazner how should we test this so we can merge for 0.7? We don't have CUDA CI yet, but I think we should have multiple people test locally to verify before merging

**Comment by pazner** [comment]
> @jedbrown @pazner how should we test this so we can merge for 0.7? We don't have CUDA CI yet, but I think we should have multiple people test locally to verify before merging

I can test locally. What exactly are the tests that should be run (is `make prove-all` enough)?

**Comment by jeremylt** [comment]
Well, how can we test that this interface is correctly handing a CUfunction off to the CUDA backend and that backend can use the CUfunction while applying the action of a QFunction?

**Comment by pazner** [comment]
It seems a bit tricky since as far as I know there is nothing in libCEED currently that generates a CUfunction that we can pass in...

**Comment by jedbrown** [comment]
Can we have a CUDA test (only run it when CUDA is available) that generates one and uses it?

File: interface/ceed-qfunction.c
Comment (jedbrown) [inline interface/ceed-qfunction.c:677]: Rather than having globally-defined macros `CEED_CUDA_ENABLED`, could we just move these functions to separate files that are only built whin CUDA (HIP, respectively) are available? That saves anti-patterns of checking those macros and avoids needing to force rebuild if adding/removing them from the configuration.

Make doesn't know when variables change, so if you
```console
make      # builds without CUDA
make CUDA_DIR=/path/to/cuda # builds CUDA sources
```
with the macros, this file won't be recompiled, leading to an inconsistent build. There are ways to ensure that make rebuilds in such conditions, but they are a blunt instrument and simpler is better when possible.

Comment (jeremylt) [reply interface/ceed-qfunction.c:647]: I added a quick fix but didn't test. Headed to appointment, but can check later if CI issue.

Comment (jeremylt) [reply interface/ceed-qfunction.c:647]: note to self, need to drop the defs of CEED_*_ENABLED

Comment (jeremylt) [reply interface/ceed-qfunction.c]: Ok, rebased off main and fixed it up. It builds fine locally for CUDA and HIP

**Review by jedbrown — APPROVED** [review]

**Comment by jedbrown** [comment]
"Decide how to test" is still "ToDo" at the top of this page so I'll let you decide if it's done.

I don't understand the GitHub policy change that one apparently needs administrator privileges to merge when a branch is "out-of-date" in a conflict-free way. It feels like this was a behavior change not too long ago, but perhaps it's a changed setting. Anyway, I can do an admin merge when ready.

**Comment by jeremylt** [comment]
You can merge now. I plan on doing a testing cleanup after 0.7

**Pull request merged**
